src/yarp.c

```diff
@@ -2241,10 +2244,65 @@ parse_expression_prefix(yp_parser_t *parser) {
       yp_node_t *statements = parse_statements(parser, YP_CONTEXT_BEGIN);
 
       accept_any(parser, 2, YP_TOKEN_NEWLINE, YP_TOKEN_SEMICOLON);
-      expect(parser, YP_TOKEN_KEYWORD_END, "Expected `end` to close `begin` statement.");
 
-      yp_token_t end_keyword = parser->previous;
-      return yp_node_begin_node_create(parser, &begin_keyword, statements, &end_keyword);
+      yp_token_t end_keyword;
+      not_provided(&end_keyword, parser->previous.end);
+
+      yp_node_t *parent = yp_node_begin_node_create(parser, &begin_keyword, statements, NULL, &end_keyword);
+
+      yp_node_t *current = NULL;
+
+      // Parse any number of rescue clauses. This will form a linked list of if
+      // nodes pointing to each other from the top.
+      while (parser->current.type == YP_TOKEN_KEYWORD_RESCUE) {
+        parser_lex(parser);
+        yp_token_t rescue_keyword = parser->previous;
+
+        yp_token_t exception_variable;
+        not_provided(&exception_variable, parser->previous.end);
+
+        yp_node_t *statements = yp_node_statements_create(parser);
+        yp_node_t *rescue = yp_node_rescue_node_create(parser, &rescue_keyword, &exception_variable, statements, NULL);
+
+        while (parser->current.type == YP_TOKEN_CONSTANT) {
+          yp_node_t *expression = parse_expression(parser, BINDING_POWER_NONE, "Expected to find a class.");
+            yp_node_list_append(parser, rescue, &rescue->as.rescue_node.exception_classes, expression);
+
+          if (accept(parser, YP_TOKEN_NEWLINE) || accept(parser, YP_TOKEN_EQUAL_GREATER)) break;
+          expect(parser, YP_TOKEN_COMMA, "Expected an ',' to delimit exception classes.");
+        }
+
+        if (parser->previous.type == YP_TOKEN_EQUAL_GREATER) {
+          expect(parser, YP_TOKEN_IDENTIFIER, "Expected variable name after `=>` in rescue statement");
+          rescue->as.rescue_node.exception_variable = parser->previous;
+        }
+
+        accept_any(parser, 2, YP_TOKEN_NEWLINE, YP_TOKEN_SEMICOLON);
+
+        rescue->as.rescue_node.statements = parse_statements(parser, YP_CONTEXT_RESCUE);
+        accept_any(parser, 2, YP_TOKEN_NEWLINE, YP_TOKEN_SEMICOLON);
+
+        if (current == NULL) {
+          parent->as.begin_node.rescue_node = rescue;
+        } else {
+          current->as.rescue_node.consequent = rescue;
+        }
+        current = rescue;
+      }
+
+      switch (parser->current.type) {
+        case YP_TOKEN_KEYWORD_END: {
+          parser_lex(parser);
+          parent->as.begin_node.end_keyword = parser->previous;
+          break;
+        }
+        default:
+          expect(parser, YP_TOKEN_KEYWORD_END, "Expected `end` to close `begin` statement.");
```

Comment:
You can just use `expect` here, as that will effectively do what you're doing in this switch